### ISSUE TITLE
size_estimates table is way off in 2.3

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#3916]
Two users have complained that the size_estimates table is producing bogus information. Spark uses that table to determine how much to read in a single pass, so problems in this table translate into spark having bad results and/or memory issues from reading too much.

Although I don't yet know more about the nature of their issues, I looked at the size_estimates with one of the users and the partition sizes seem to be too small to be true - around ~500 bytes.

Looking at the code, I see that we are deriving that from `sstable->get_stats_metadata().estimated_row_size`

I remember, from looking at the large partition table, that at some point in the past that would give us information about a partition, and more recently it started exporting information about a single individual row. If that is the case, my hunch is that we are exporting the average size of a row as if it were the average size of an entire partition.

Specially for users with larger partitions, that would indeed generate very large reads from spark.

@psarna @duarten since both of you have touched this in the past - does that make sense ?

**Comment by glommer** [comment]
This is strange: it seems to me this code is still updating histograms for partition sizes:

```
        add_row_size(stats.partition_size);
```

`stats.partition_size` is the same metric that we are using in the large partition detector.

```  
void add_row_size(uint64_t row_size) {
        _estimated_row_size.add(row_size);
    }
```

The above seems to correctly update `_estimated_row_size`. 

```
    void construct_stats(stats_metadata& m) {
        m.estimated_row_size = std::move(_estimated_row_size);
```

The above means that we are adding the estimated_row_size correctly to the estimates. That is the statistic that we use to create the size_estimates table. Yet, it is, in practice, way lower than it should.

**Comment by glommer** [comment]
One of the users was kind enough to provide us a snippet of 2.2 and then 2.3 right after

2.2:

```
keyspace_name   | table_name | range_start          | range_end            | mean_partition_size | partitions_count
-----------------+------------+----------------------+----------------------+---------------------+------------------
 augury_research |   machines | -1004397762444347106 |  -852463915728168461 |                5976 |                1
 augury_research |   machines | -1079232473097286959 | -1004397762444347106 |                5976 |                1
 augury_research |   machines |  -112550900330953082 |    16868306819291740 |                5976 |                1
```

2.3:

```
keyspace_name   | table_name | range_start          | range_end            | mean_partition_size | partitions_count
-----------------+------------+----------------------+----------------------+---------------------+------------------
 augury_research |   machines | -1004397762444347106 |  -852463915728168461 |         95906433959 |                1
 augury_research |   machines | -1079232473097286959 | -1004397762444347106 |         95906433959 |                1
 augury_research |   machines |  -112550900330953082 |    16868306819291740 |         95906433959 |                1
 augury_research |   machines | -1143243528061381118 | -1079232473097286959 |         95906433959 |                1
 augury_research |   machines | -1241041063772061064 | -1143243528061381118 |         95906433959 |                1
```

**Comment by duarten** [comment]
I wrote the original implementation but haven't touched it much since. We could.. bisect :/

**Comment by psarna** [comment]
As for large partitions table, it was (and still is) in partition granularity, not rows - it's updated via maybe_update_large_partitions() that takes partition size as one of the arguments.

I made some simple local tests on master(447f953a2c0d42a4a0bae833135b6d395e1fa99e) and data from system.size_estimates was consistent with both system.large_partitions and partition size on disk... So even before bisecting we'd need to reproduce - any clues what kind of load causes this inconsistencies?

**Comment by glommer** [comment]
reproduced using the following script:

```
#!/usr/bin/env python

import sys

from datetime import datetime, timedelta
from cassandra import OperationTimedOut, WriteTimeout, Unavailable
from cassandra.cluster import Cluster
from cassandra import ConsistencyLevel, policies 
from cassandra.query import dict_factory, BatchStatement, BatchType, SimpleStatement
from random import choice, randint
from uuid import uuid4
from time import sleep

futures = []


def create_keyspace(keyspace, session):
    session.execute("""CREATE KEYSPACE IF NOT EXISTS %s
            WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1}
            AND DURABLE_WRITES = true""" % keyspace)

cluster = Cluster(
    ['172.30.0.242'], executor_threads=10
)

session = cluster.connect()
session.default_consistency_level = ConsistencyLevel.LOCAL_QUORUM
session.row_factory = dict_factory

create_keyspace('repro', session)
session.set_keyspace('repro')

session.execute("DROP TABLE IF EXISTS repro;")

session.execute("""CREATE TABLE IF NOT EXISTS repro(
    pk int,
    ck int,
    val blob,
    PRIMARY KEY (pk, ck, val)
    )
    WITH CLUSTERING ORDER BY (ck DESC)
    """)

prepared = session.prepare(
    "INSERT INTO repro (pk, ck, val) VALUES (?, ?, ?)")


coordinator = cluster.get_control_connection_host()
print('coordinator: %s' % coordinator)

for pk in range(100):
    for ck in range(100):
        try:
            future = session.execute_async(prepared, [pk, ck, 0xdeadbeef])
            futures.append(future)
        except KeyboardInterrupt:
            break

# synchrounous wait for the remaining queries to finish
for f in futures:
    try:
        f.result()
    except Exception as err:
        failed_callback(err)

cluster.shutdown()
```

After the upgrade to 2.3, size estimates get completely wrong. However, after I call `nodetool compact`, they get back to sanity.

this could be a la vs ka thing, since 2.3 is the version where we cross the ka -> la boundary.

@azk does that work for you if you call `nodetool compact augury_research` ?

**Comment by avikivity** [comment]
Another possible cause is just a restart. It's possible that reading statistics from disk is broken, but having the statistics being generated (by memtable flush or compaction) makes them correct again.

@glommer does a simple restart, without an update, also breate size_estimates?

**Comment by glommer** [comment]
No, a restart between 2.2 and 2.2 doesn't cause this. It survives a restart just fine.

My suspicion is that we are not parsing old formats correctly for stats , just the current format.

**Comment by azk** [comment]
@glommer it worked!

It returned the `size_estimates` to its original size on my one node test setup and I'm also able to run spark queries against the table.

I'm now running compact in our cluster. Thank you!

**Comment by glommer** [comment]
@avikivity ok, so full set of tests:

1) inserted on 2.2 -> sizes are ok
2) restarted -> sizes are ok
3) upgraded to 2.3 -> sizes are bogus
4) restarted -> sizes are still bogus (unchanged)
5) compacted -> sizes are okay.

So definitely the issue here is in the ka -> la transition

**Comment by glommer** [comment]
the plot thickens.

If I restart a 2.3 node, then the sizes get to be bogus again. So both things are happening. I though to check that after I saw these patches:

```
commit df457166b0ee58cd19f6d347a0059f0f382264b6
Author: Piotr Jastrzebski <piotr@scylladb.com>
Date:   Thu Mar 29 16:18:00 2018 +0200

    Add support for 3_x stats metadata
    
    Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>

commit e1e23ec555c301fe4bac85d38ed2a31a19177844
Author: Piotr Jastrzebski <piotr@scylladb.com>
Date:   Thu Mar 29 15:34:11 2018 +0200

    Pass sstable version to describe_type
    
    Signed-off-by: Piotr Jastrzebski <piotr@scylladb.com>
```

So Maybe on 2.2 everything works fine (restart or no restart), but on 2.3 it only works for Stats that we already have in memory.

@haaawk comments ?

**Comment by glommer** [comment]
The histograms are busted.

Here's what happens with the histogram that is read from disk : 

```
...
[17436918..20924300]: 0                                                                        
[20924301..25109160]: 0                                                                        
[25109161..1]: 0
[2..1]: 0
[2..2]: 0
...
```

The offsets are non-monotonic and are reset back to 0. And it is already like that when it comes from disk.
That doesn't happen with the histogram post compaction.

**Comment by glommer** [comment]
This is likely my fault and seem to have been broken by a33f0d6f92360ab94688e7525ab9ff4dcdaf514a

Still trying to understand exactly why.

**Comment by glommer** [comment]
Patch posted.

**Comment by glommer** [comment]
Fixed by e13e7962906af5bc1a0a7ccc876571f615adc30b.

Got the wrong issue number in the patch and it didn't auto-close
